### PR TITLE
fix: Use Q8_0 for all embedding quantizations for granite and granitemoe

### PR DESCRIPTION
**Pull request by gabe-l-hart**
## Description

At lower precision levels, the models can manifest numerical instability, especially with batch size > 1. This shows up as nondeterministic stopping when index 0 (the EOG token) has a seemingly uninitialized large value in the logits.

**NOTE**: The same behavior can be achieved by adding `--token-embedding-type Q8_0` to `llama-quantize`. The goal of this PR is to avoid needing to document this for all users that are performing their own quantization of the Granite models.

## Changes

* Add checks in `llama-quant.cpp` to always use `GGML_TYPE_Q8_0` for token embeddings.
    * The checks are in both the clause for tied embeddings and non-tied embeddings since this is a configurable parameter for these architectures

## Testing

This issue came up around the published Ollama models for `granite3.1-dense` and `granite3.1-moe`. The instability (early stopping) arose only when running multiple requests in parallel that resulted in a batch with size > 1. I used the script below to repro against both Ollama and the native `llama-server`:

<details>
<summary>parallel_repro.py</summary>

```py
#!/usr/bin/env python

import argparse
import concurrent.futures
import requests
from functools import partial


prompts = [
    "Hello world! Tell me about yourself",
    "Tell me a story about a developer and their dog",
    "Write a poem about the beauty of nature",
    "Write an email to your future self",
    "Tell me about your dreams and aspirations",
    "Write a letter to your parents",
    "Write a short story about a detective solving a murder case",
    "Write a script for a short film",
    "Write a blog post about the importance of mental health",
    "Write an essay about the history of technology",
]

def do_chat_ollama(args, prompt):
    body = {
        "model": args.model,
        "stream": False,
        "messages": [
            {
                "role": "user",
                "content": prompt,
            },
        ],
        "options": {
            "temperature": 0,
            "num_predict": args.max_t
            # "num_gpu": 39,
            # "cache_prompt": not no_cache,
        },
    }
    response = requests.post(f"{args.base_url}/api/chat", json=body)
    output = None
    done_reason = None
    if response.status_code == 200:
        j = response.json()
        output = j["message"]["content"]
        done_reason = j["done_reason"]
    else:
        print("ERROR")
        print(response.text)
    return (prompt, {"response": output, "done_reason": done_reason})


def do_chat_llamacpp(args, prompt):
    body = {
        "model": args.model,
        "stream": False,
        "messages": [
            {
                "role": "user",
                "content": prompt,
            },
        ],
        "temperature": 0,
        "max_tokens": args.max_tokens,
    }
    response = requests.post(f"{args.base_url}/v1/chat/completions", json=body)
    output = None
    done_reason = None
    if response.status_code == 200:
        j = response.json()
        output = j["choices"][0]["message"]["content"]
        done_reason = j["choices"][0]["finish_reason"]
    else:
        print("ERROR")
        print(response.text)
    return (prompt, {"response": output, "done_reason": done_reason})


def do_generate_ollama(args, prompt):
    body = {
        "model": args.model,
        "stream": False,
        "prompt": prompt,
        "options": {
            "temperature": 0,
            # "cache_prompt": not args.no_cache,
            "num_predict": args.max_tokens,
        },
    }
    response = requests.post(f"{args.base_url}/api/generate", json=body)
    output = None
    done_reason = None
    if response.status_code == 200:
        j = response.json()
        output = j["response"]
        done_reason = j["done_reason"]
    else:
        print("ERROR")
        print(response.text)
    return (prompt, {"response": output, "done_reason": done_reason})


def do_generate_llamacpp(args, prompt):
    body = {
        "model": args.model,
        "stream": False,
        "prompt": prompt,
        "temperature": 0,
        "max_tokens": args.max_tokens,
    }
    response = requests.post(f"{args.base_url}/v1/completions", json=body)
    output = None
    done_reason = None
    if response.status_code == 200:
        j = response.json()
        output = j["choices"][0]["text"]
        done_reason = j["choices"][0]["finish_reason"]
    else:
        print("ERROR")
        print(response.text)
    return (prompt, {"response": output, "done_reason": done_reason})

if __name__ == "__main__":

    parser = argparse.ArgumentParser()
    parser.add_argument("-j", "--num-threads", type=int, default=4, help="Number of concurrent threads")
    parser.add_argument("-M", "--mode", choices=["generate", "chat"], default="chat", help="Run as raw generation or chat")
    parser.add_argument("-n", "--num-prompts", type=int, default=len(prompts), help="Number of prompts to send")
    parser.add_argument("-u", "--base-url", default="http://localhost:11434", help="Base URL of the server")
    parser.add_argument("-m", "--model", default="granite3.1-dense", help="The Ollama model to run")
    parser.add_argument("-c", "--no-cache", action="store_true", help="Run with no caching")
    parser.add_argument("-s", "--server-type", choices=["ollama", "llama.cpp"], default="ollama", help="Which type of server to test against")
    parser.add_argument("-x", "--max-tokens", type=int, default=100, help="Max tokens to generate")
    args = parser.parse_args()

    if args.server_type == "ollama":
        mode_fn = do_chat_ollama if args.mode == "chat" else do_generate_ollama
    else:
        mode_fn = do_chat_llamacpp if args.mode == "chat" else do_generate_llamacpp
    doit = partial(mode_fn, args)

    with concurrent.futures.ThreadPoolExecutor(max_workers=args.num_threads) as executor:
        results = dict(executor.map(doit, prompts[:args.num_prompts]))

    for prompt, response in sorted(results.items()):
        print("------------------------")
        print("PROMPT")
        print(prompt)
        print()
        print("RESPONSE")
        print(response["response"])
        print()
        print("DONE REASON")
        print(response["done_reason"])
        print()

```
</details>

Using this script, I did the following:

```sh
# (Terminal 1) Run the model with `llama-server`
./build/bin/llama-server --model ggml-model-Q4_K_M.gguf --parallel 4 --batch-size 512 --port 8081

# (Terminal 2) Run the parallel script with -j 1 to verify expected outputs
python parallel_repro.py -u http://localhost:8081 -j 1 -n 3 -s llama.cpp -x 300

# (Terminal 2) Run the parallel script with -j 3 to check for truncated responses
python parallel_repro.py -u http://localhost:8081 -j 3 -n 3 -s llama.cpp -x 300
```

Using this, I was able to verify that without this fix, and using standard `llama-quantize <path/to/f16.gguf> Q4_K_M` reproduced the early stopping. Without this fix, but with `--token-embedding-type Q8_0`, the results were no longer truncated. With this fix, I no longer needed to add the special conversion argument.

While investigating, I tested this against a large number of out-of-the-box quantizations from Ollama ([here](https://ollama.com/library/granite3-dense/tags)) across both `granite` and `granitemoe`. I found that for both `granite` and `granitemoe`, the early stopping occurs across all quantizations below `Q8_0`. With this fix, I validated that early stopping is gone for both `granite` and `granitemoe` at `Q4_K_M`.

### CODE REVIEW

**Comment by arch-btw** [comment]
Thank you, do you happen to know if the original implementation (or this pull request) affects `--token-embedding-type f16`?
I'm curious whether it has to be exactly Q8_0 or if higher is also functional. 👍

**Comment by gabe-l-hart** [comment]
@arch-btw That's a good question that I should definitely answer before merging. My assumption is that the clause above takes care of this, but I should definitely validate that:

```c++
        if (qs.params->output_tensor_type < GGML_TYPE_COUNT) {
            new_type = qs.params->output_tensor_type;
        } else {
```

**Comment by gabe-l-hart** [comment]
Reading further, this clause does _not_ do what I had originally thought. It appears to be the equivalent of checking whether the `--output-tensor-type` arg was sent explicitly, so I think it's probably worth an explicit check to ensure that the target quantization is smaller than `Q8` before increasing to `Q8`.

**Comment by gabe-l-hart** [comment]
Fix pushed! I used `ggml_is_quantized` and only flatten to `Q8_0` if true. Thanks for catching this

**Comment by ggerganov** [comment]
The root cause is actually a bug in the mat-vec Metal kernels. They are doing out-of-bounds write when the `src0` rows are not multiples of the number of launched threadgroups. For example, you can fix the `Q4_K` model with the original embeddings using the following patch:

```diff
diff --git a/ggml/src/ggml-metal/ggml-metal.metal b/ggml/src/ggml-metal/ggml-metal.metal
index 8ba43904d..3203170d5 100644
--- a/ggml/src/ggml-metal/ggml-metal.metal
+++ b/ggml/src/ggml-metal/ggml-metal.metal
@@ -4906,6 +4906,10 @@ void kernel_mul_mv_q6_K_f32_impl(
 
     const int row = 2*r0 + sgitg;
 
+    if (row >= args.ne0) {
+        return;
+    }
+
     const uint i12 = im%args.ne12;
     const uint i13 = im/args.ne12;
 

```

Similar patch has to be applied to the rest of the mat-vec kernels.

@gabe-l-hart Please confirm that this fixes the issue and I will make a PR to fix these.

**Comment by gabe-l-hart** [comment]
Thank you! This makes so much more sense with what I was seeing. I'll test your fix on Monday and let your know the results.

**Comment by gabe-l-hart** [comment]
@ggerganov Confirmed! That does the trick with the original `Q4_K_M` model where the `token.embd.weight` tensor is quantized to `q6_K`.
